sql/opt_range.cc

```diff
@@ -3343,11 +3360,43 @@ bool calculate_cond_selectivity_for_table(THD *thd, TABLE *table, Item **cond)
           table->quick_key_parts[keynr] == quick_key_parts)
       {
         uint i;
-        uint used_key_parts= table->quick_key_parts[keynr];
         double quick_cond_selectivity= table->quick_rows[keynr] / 
 	                               table_records;
         KEY *key_info= table->key_info + keynr;
         KEY_PART_INFO* key_part= key_info->key_part;
+
+        /*
+          If we use the extended keys, some of the range conditions
+          on primary keys will get shadowed and will not be considered and
+          can cause performance degradations due to selecting less performing
+          plans. Therefore, the following code is changed to use only user defined
+          key parts (without primary key extensions)
+         */
+        uint keynr_quick_key_parts= table->quick_key_parts[keynr];
+
+        /*
+          If the current key part is from an extended key, then we need to skip
+          it.
+        */
+        if ((key_info->ext_key_parts > 0) && // current key_info has extended key parts
+            (pk_info != NULL))
+        {
+          // reset extended key column buffer
+          bitmap_clear_all(&extended_key_columns);
+
+          // extended key parts exist
+          uint k;
+          for(k=0; k < pk_info->usable_key_parts; ++k)
+          {
+            if (key_info->ext_key_part_map & (1 << k))
+            {
+              KEY_PART_INFO* pk_key_part= pk_info->key_part + k;
+              // mark column as used.
+              bitmap_set_bit(&extended_key_columns, pk_key_part->fieldnr-1);
+            }
```

Comment:
So, this is populating `extended_key_columns`  which is a bitmap of columns that are A) present in the primary key and B) present in key `keynr`.   Then, below,    `extended_key_columns` is used to make the checks against the key parts of this index.

A question: why can't we populate `extended_key_columns` once, putting all the columns into it?

Reply:
.. no we can't because `key_info->ext_key_part_map` is different for every key.

But then, another question arises:  so, with this patch, the optimizer behavior for 
```
key(col1), primary key(pk1, pk2)
```
will be different from 
```
key(col1, pk1), primary key(pk1, pk2)
```
doesn't this defeat the purpose of Extended Keys feature  (whose goal is to make the optimizer fully aware of key extensions, that is, the optimizer should behave as if every key had all pk columns explicitly declared for it)  ?